tests/core/contracts/test_contract_call_interface.py

```diff
@@ -1149,6 +1152,15 @@ def test_changing_default_block_identifier(w3, math_contract):
     assert math_contract.functions.counter().call(block_identifier=None) == 7
 
 
+def test_functions_iterator(w3, math_contract):
+    all_functions = math_contract.all_functions()
+    functions_iter = math_contract.functions
+
+    for fn, expected_fn in zip(iter(functions_iter), all_functions):
+        assert isinstance(fn, BaseContractFunction)
```

Comment:
Want to check this type is correct, or is there a better one I may have missed?

Reply:
Since the tests you have are for sync and async providers, not just an instance of BaseProvider, you could `assert isinstance(fn, ContractFunction)` here and `AsyncContractFunction` below and keep both tests. Not sure it's necessary, but checks that the class inherits properly from base :shrug: